parsl/data_provider/zip.py

```diff
@@ -93,6 +113,18 @@ def _zip_stage_out_app(dm):
     return parsl.python_app(executors=['_parsl_internal'], data_flow_kernel=dm.dfk)(_zip_stage_out)
 
 
+def _zip_stage_in(zip_file, inside_path, working_dir, *, parent_fut, outputs, _parsl_staging_inhibit=True):
+    with filelock.FileLock(zip_file + ".lock"):
+        with zipfile.ZipFile(zip_file, mode='r') as z:
+            content = z.read(inside_path)
+        with open(outputs[0], "wb") as of:
+            of.write(content)
```

Comment:
I wonder if there's a way to think in lists.  This method:
1. takes out a lock
2. opens the (zip) file
3. reads the content from the zip file
4. closes the (zip) file
5. opens a new file
6. writes the content
7. closes the new file

There's clearly no working around steps 3, 5, 6, and 7, but steps 1, 2, and 4 could be aggregated for multiple reads, if, say, `inside_path` were of type `List` and there were an appropriate number of outputs.

That could further be better for memory by reading and writing in chunks.  Consider:
```python
with filelock.FileLock(...):
    with zipfile.ZipFile(...):
        for inp, outp in zip(inside_paths, outputs):
            with open(outp, "wb") as of:
                shutil.copyfileobj(inp, outp)
```

I won't write a similar comment for `_zip_stage_out`, but I'm thinking in similar terms.  "It would be nice if this could be implemented with list-think."